File: app/src/main/java/edu/pokemon/iut/pokedex/data/dao/PokemonDao.java
package edu.pokemon.iut.pokedex.data.dao;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.Query;
import android.arch.persistence.room.Update;

import java.util.List;

import edu.pokemon.iut.pokedex.data.model.Pokemon;
import edu.pokemon.iut.pokedex.data.model.Sprites;

import static android.arch.persistence.room.OnConflictStrategy.REPLACE;

@Dao
public interface PokemonDao {
    @Insert(onConflict = REPLACE)
    void save(Pokemon pokemon);

    @Query("SELECT * FROM pokemon WHERE id = :pokemonId")
    LiveData<Pokemon> load(int pokemonId);

    @Query("SELECT * FROM pokemon")
    LiveData<List<Pokemon>> loadAll();

    @Query("SELECT * FROM pokemon WHERE name LIKE :query")
    LiveData<List<Pokemon>> loadAllWithFilter(String query);

    @Query("SELECT id from pokemon where id = :pokemonId")
    int hasPokemon(int pokemonId);

    @Query("SELECT COUNT(*) FROM pokemon")
    LiveData<Integer> getNumberOfPokemon();

    @Update(onConflict = REPLACE)
    void capture(Pokemon pokemon);

}

File: app/src/main/java/edu/pokemon/iut/pokedex/data/dao/PokemonDao.kt
package edu.pokemon.iut.pokedex.data.dao

import android.arch.lifecycle.LiveData
import android.arch.persistence.room.Dao
import android.arch.persistence.room.Insert
import android.arch.persistence.room.Query
import android.arch.persistence.room.Update

import edu.pokemon.iut.pokedex.data.model.Pokemon

import android.arch.persistence.room.OnConflictStrategy.REPLACE

/**
 * DAO that give access to the Pokemon Database
 */
@Dao
interface PokemonDao {

    /**
     * Query to know the number of pokemon saved in database
     *
     * @return [Integer] number of pokemon, [LiveData] allow to observe any change in database and notify the observer.
     */
    @get:Query("SELECT COUNT(*) FROM pokemon")
    val numberOfPokemon: LiveData<Int>

    /**
     * Insert a pokemon, if it exist already the method will replace it.
     *
     * @param pokemon [Pokemon] to save
     */
    @Insert(onConflict = REPLACE)
    fun save(pokemon: Pokemon)

    /**
     * Query to retrieve a specific pokemon by is ID
     *
     * @param pokemonId of the pokemon to retrieve
     * @return [Pokemon] for the given id, [LiveData] allow to observe any change in database and notify the observer.
     */
    @Query("SELECT * FROM pokemon WHERE id = :pokemonId")
    fun load(pokemonId: Int): LiveData<Pokemon>

    /**
     * Query to retrieve ALL the pokemon
     *
     * @return [<], [LiveData] allow to observe any change in database and notify the observer.
     */
    @Query("SELECT * FROM pokemon")
    fun loadAll(): LiveData<List<Pokemon>>

    /**
     * Query to retrieve SOME pokemon filtered by their name
     *
     * @param query name or portion of name
     * @return [<], [LiveData] allow to observe any change in database and notify the observer.
     */
    @Query("SELECT * FROM pokemon WHERE name LIKE :query")
    fun loadAllWithFilter(query: String): LiveData<List<Pokemon>>

    /**
     * Query to know if a pokemon already exist or not in the database
     *
     * @param pokemonId of the one to check existence
     * @return id of the pokemon
     */
    @Query("SELECT id from pokemon where id = :pokemonId")
    fun hasPokemon(pokemonId: Int): Int

    /**
     * Query to update the status of capture for a pokemon
     *
     * @param pokemon [Pokemon] to update
     */
    @Update(onConflict = REPLACE)
    fun capture(pokemon: Pokemon)

}